builder/replica_pipeline_test.cc

```diff
@@ -0,0 +1,1129 @@
+TEST_F(ReplicaPipelineTest, matmul) {
+  constexpr int split = 1;
+  constexpr loop_mode lm = loop_mode::serial;
+
+  // Make the pipeline
+  node_context ctx;
+
+  auto a = buffer_expr::make(ctx, "a", sizeof(int), 2);
+  auto b = buffer_expr::make(ctx, "b", sizeof(int), 2);
+  auto c = buffer_expr::make(ctx, "c", sizeof(int), 2);
+  auto abc = buffer_expr::make(ctx, "abc", sizeof(int), 2);
+  auto ab = buffer_expr::make(ctx, "ab", sizeof(int), 2);
+
+  var i(ctx, "i");
+  var j(ctx, "j");
+
+  // The bounds required of the dimensions consumed by the reduction depend on the size of the
+  // buffers passed in. Note that we haven't used any constants yet.
+  auto K_ab = a->dim(1).bounds;
+  auto K_abc = c->dim(0).bounds;
+
+  // We use int for this pipeline so we can test for correctness exactly.
+  func matmul_ab = func::make(matmul<int>, {{a, {point(i), K_ab}}, {b, {K_ab, point(j)}}}, {{ab, {i, j}}});
+  func matmul_abc = func::make(matmul<int>, {{ab, {point(i), K_abc}}, {c, {K_abc, point(j)}}}, {{abc, {i, j}}});
+
+  a->dim(1).stride = a->elem_size();
+  b->dim(1).stride = b->elem_size();
+  c->dim(1).stride = c->elem_size();
+  abc->dim(1).stride = abc->elem_size();
+
+  // TODO: There should be a more user friendly way to control the strides.
+  ab->dim(1).stride = static_cast<index_t>(sizeof(int));
+  ab->dim(0).stride = ab->dim(1).extent() * ab->dim(1).stride;
+
+  if (split > 0) {
+    matmul_abc.loops({{i, split, lm}});
+
+    if (lm == loop_mode::parallel) {
+      ab->store_at({&matmul_abc, i});
+    }
+  }
+
+  pipeline p = build_pipeline(ctx, {a, b, c}, {abc});
+  pipeline p_replica = kMatmulReplica();
+
+  // Look at the source code to this test to verify that we
+  // we have something that matches exactly
+  std::string replica_text = define_replica_pipeline(ctx, {a, b, c}, {abc});
+  LOG_REPLICA_TEXT(replica_text);
+  size_t pos = replica_pipeline_test_src.find(replica_text);
+  ASSERT_NE(pos, std::string::npos) << "Matching replica text not found, expected:\n" << replica_text;
+
+  // Run the pipeline
+  const int M = 10;
+  const int N = 10;
+  buffer<int, 2> a_buf({N, M});
+  buffer<int, 2> b_buf({N, M});
+  buffer<int, 2> c_buf({N, M});
+  buffer<int, 2> abc_buf({N, M});
+  // TODO: There should be a more user friendly way to initialize a buffer with strides other than the default
+  // order.
+  std::swap(a_buf.dim(1), a_buf.dim(0));
+  std::swap(b_buf.dim(1), b_buf.dim(0));
+  std::swap(c_buf.dim(1), c_buf.dim(0));
+  std::swap(abc_buf.dim(1), abc_buf.dim(0));
+
+  init_random(a_buf);
+  init_random(b_buf);
+  init_random(c_buf);
+  abc_buf.allocate();
+
+  const raw_buffer* inputs[] = {&a_buf, &b_buf, &c_buf};
+  const raw_buffer* outputs[] = {&abc_buf};
+  {
+    test_context eval_ctx;
+    p.evaluate(inputs, outputs, eval_ctx);
+  }
+  {
+    test_context eval_ctx;
+    p_replica.evaluate(inputs, outputs, eval_ctx);
+  }
+}
+
+// clang-format off
+static std::function<pipeline()> kPyramidReplica =
```

Comment:
A lot of the naming style of things here is "google style" which is inconsistent. The style I've been using is trying to just match the C++ standard library, which is basically lowercase everything with underscores, and CapitalWords for template parameters AFAICT.

Reply:
We don't have an explicit style guide or a .clang-tidy to enforce these things, so they'll slip through the cracks. (I wrote constant data as `kFoo` for about 30+ years now when not enforced to do otherwise so old habits die hard.) Sounds like you want `snake_case` for these too, so, done.